### ISSUE TITLE
Multi-part upload completion very slow and eventually timed out for >=9GB files

### ISSUE THREAD

**Issue by henryzzhang** [minio/minio#3223]
This issue might be related to #2334, which seems has been closed.  But I got similar issue when I am using minio server built from source code (downloaded 11/09/2016).  Basically, I was uploading a 9GB file divided into 100 chunks.  After all chunks have been uploaded successfully with multiple worker threads, I sent the multi-part upload completion request, which blocks for a long time and then timed out.

## Expected Behavior

If issue #2334 was fixed, the completion request should return much faster because the concatenation was performed during chunks uploading phase.  With aws s3, no such delay or timeout was observed for huge file upload completion.  Theirs appears to be an async call.

## Current Behavior

Long-time blocking call for the multi-part upload completion and eventually timed out for huge files.

## Possible Solution

Make sure the concatenation was performed while parts where being uploaded or make the completion request become async.  Concatenation during uploading might not be achievable if the client sends lots of chunks in parallel.  Making the completion call async may be a better option.

## Steps to Reproduce (for bugs)

1.Create a huge file, say 10GB
2.Get multi-part pre-signed urls (100 parts)
3.Upload all parts using 4 worker threads (sending 4 parts at a time in parallel)
4.Send completion request after all chunks have been uploaded
5. You will see the delay and timeout eventually

## Context

I am trying to speed up the uploading of huge files using multiple worker threads for the chunks.

## Your Environment

* Version used:
minio server built from source code downloaded on 11/09/2016

* Environment name and version (e.g. nginx 1.9.1):
No nginx used.  Both client and minio server are in the same local network for testing.

* Server type and version:
latest version 11/09/2016

* Operating System and version:
Ubuntu 16.04

* Link to your project:

**Comment by harshavardhana** [comment]
Multipart uploads are concatenated on a regular FS backend - so you need to wait for CompleteMultipartUpload(). But this concatenation is optimal when sequence of upload uploads 1st part and so .. it is sub-optimal if you upload 100th part first and 1st part last. 

FS needs to keep the namespace consistent so it allows this behavior so this is not an issue per se but a behavior we wanted to keep and will keep into near future.  

FWIW - concatenation doesn't happen on erasure coded backend.

**Comment by henryzzhang** [comment]
I did wait to let CompleteMultipartUpload() to finish, but it failed with
http timeout error.  After that, I found the file has actually been
assembled correctly in the bucket. But since my client code get failure
from CompleteMultipartUpload, it thought the file was not uploaded
successfully.

The failure happens when I use 4 worker threads to send the chunks in
parallel.  However, if I just use one thread to upload the chunks (one
after another in sequence), I don't get the timeout.

Just as they did in aws s3, I suggest you make CompleteMultipartUpload()
async and use a background thread to start or finish the concatenation.
The client doesn't need to wait for the file to become ready.  This is
typical in distributed computing environment.  With your current
implementation in minio, timeout will still occur even if I upload chunks
in sequence if each chunk is too large.

**Comment by harshavardhana** [comment]
> Just as they did in aws s3, I suggest you make CompleteMultipartUpload()
> async and use a background thread to start or finish the concatenation.
> The client doesn't need to wait for the file to become ready.  This is
> typical in distributed computing environment.  With your current
> implementation in minio, timeout will still occur even if I upload chunks
> in sequence if each chunk is too large.

We had this before but we removed so that to not complicate the codebase just for FS backend alone, If you use erasure coded backend then the CompleteMultipartUpload is instantaneous since all parts are kept as is with the support of the backend format.  It is even write/read consistent as soon as the writes are completed you can read as well.  Due to special requirements of FS so that as a user when you feel like not using minio you should be able to happily use the backend and copy data out with regular tools. The benefits this brings in outweighed the one off timeouts that can occur and it is also when parts are uploaded in a un-ordered fashion.

Lets see from the code point of view itself. Now with 4 workers how are the parts uploaded is the first part uploaded as part of the first 4 workers?  or does it get uploaded later ? do you have control over which parts go first?

**Comment by jboelter** [comment]
Would you consider a mode of operation where concatenation is not performed on a FS backend? It would seem that the concatenation feature is only useful if you are accessing the files via the filesystem vs only the REST api.

**Comment by harshavardhana** [comment]
> Would you consider a mode of operation where concatenation is not performed on a FS backend? It would seem that the concatenation feature is only useful if you are accessing the files via the filesystem vs only the REST api.

Erasure coded backend already does this never concatenates and keeps the parts in a special format,  FS purpose was to keep the backend simple and not introduce any custom format.  Having multiple modes in FS can cause confusion and it would also discourage users to use FS. 

In this case we need understand the code which @henryzzhang used to reproduce timeouts. With FS from what we have seen and tested using mc,  aws cli, cloud berry backup all these uploads in parallel especially `aws cli` uploads 10 parts in parallel.  We don't see any timeouts at all.

**Comment by krishnasrinivas** [comment]
@henryzzhang can you post the client program source that uploads?

**Comment by henryzzhang** [comment]
I have used 4 worker threads to upload 100 chunks (chunk size = 90MB).  The
order of the 4 chunks started by the 4 threads are in-deterministic, but
the 4 chunks will be finished uploading before the next 4 chunks gets
started uploading.  The timeout will also happen with only one worker
thread if the chunk size is large enough.  I can reproduce the timeout with
one worker thread uploading 100 chunks with chunk size = 200MB.

Erasure coded backend requires multiple hard disks.  It is probably used
for mission-critical data storage.  Our usage case is different.

**Comment by henryzzhang** [comment]
My code is scattered in multiple places in different components, so it is
hard to put them all together as a working code.  My client code is
actually c++ on windows.  It uses aws s3 REST APIs.
But with all the info and steps that I have provided, I am sure you should
be able to reproduce.

**Comment by krishnasrinivas** [comment]
> The order of the 4 chunks started by the 4 threads are in-deterministic,

Ah, if the are deterministic, i.e if the 4 parallel chunks are from start to end, then it would work fine without timeout. aws-cli also uploads it in 10 parts in parallel in a batch, but still the batches itself are sequential.

**Comment by henryzzhang** [comment]
But you probably shouldn't require client to do the uploads in a
deterministic way.  Even if uploading of the 4 chunks started from 1st to
the 4th, they may finish in random order depending on the network bandwidth
over time, right?
Also, as I have mentioned, the timeout occurs even with only one worker
thread if the chunk size is large enough.

**Comment by harshavardhana** [comment]
> But you probably shouldn't require client to do the uploads in a
> deterministic way.  Even if uploading of the 4 chunks started from 1st to
> the 4th, they may finish in random order depending on the network bandwidth
> over time, right?
> Also, as I have mentioned, the timeout occurs even with only one worker
> thread if the chunk size is large enough.

So i tried with uploader.go with 64MB, 100MB and 200MB with a 8GB file not able to reproduce the issue. I tried three different concurrency options i.e 

CONCURRENCY = 2, 3 and 4 

Here is the relevant code. 

``` go
package main

import (
    "log"
    "os"

    "github.com/aws/aws-sdk-go/aws"
    "github.com/aws/aws-sdk-go/aws/credentials"
    "github.com/aws/aws-sdk-go/aws/session"
    "github.com/aws/aws-sdk-go/service/s3/s3manager"
)

func main() {
    file, err := os.Open("/home/harsha/xubuntu-xenial-mate-20160528-longsleep-pine64-8GB.img")
    if err != nil {
        log.Fatal("Failed to open file", err)
    }

    creds := credentials.NewStaticCredentials("USWUXHGYZQYFYFFIT3RE", "MOJRH0mkL1IPauahWITSVvyDrQbEEIwljvmxdq03", "")
    s3Config := &aws.Config{
        Credentials:      creds,
        Endpoint:         aws.String("http://localhost:9000"),
        Region:           aws.String("us-east-1"),
        DisableSSL:       aws.Bool(true),
        S3ForcePathStyle: aws.Bool(true),
    }
    newSession := session.New(s3Config)

    uploader := s3manager.NewUploader(newSession)
    uploader.Concurrency = 4              // 4 concurrent uploads.
    uploader.PartSize = 1024 * 1024 * 200 // 200MiB
    result, err := uploader.Upload(&s3manager.UploadInput{
        Body:   file,
        Bucket: aws.String("testbucket"),
        Key:    aws.String("object"),
    })
    if err != nil {
        log.Fatalln("Failed to upload", err)
    }

    log.Println("Successfully uploaded to", result.Location)
}
```

**Comment by harshavardhana** [comment]
Can you provide more information about hardware, backend and disks - is minio server running on a SATA 7200RPM drives?.  My tests were on SSD.  @henryzzhang

**Comment by krishnasrinivas** [comment]
Sorry, my previous comment was ambiguous. Way it works on server side is "append the parts as early as possible but keep the order of part number intact in the appended file" so that when the server gets CompleteMultipartUpload in the end there is a high chance that the appended file contains all the parts and we can return success immediately.

For ex. let's say totally 50 parts need to be uploaded, you can upload 10 parts in parallel in 5 batches. The 1st batch will have parts 1-10 and these parts can be uploaded in parallel without bothering about ordering) the 2nd batch will have parts 11-20, 3rd batch will have parts 21-30 etc. This is how aws-cli uploads.

Let's say the client uploaded parts 2-50, at this point we wont be able to start append process till the 1st part is uploaded. Hence when the 1st part is uploaded in the last and CompleteMultipartUpload() is sent, we have to wait till the concatenation process is complete before responding `success` to CompleteMultipartUpload() request (during which the client can timeout). Other option is - when we get CompleteMultipartUpload() we can return success immediately and concatenate in background, but this will make things complicated - for ex. we have to handle things like server crashing when concatenation is happening when we have already returned success to the client.

The advantage for AWS S3 is they don't have to keep the parts in one appended file and instead keep the parts as different files and maintain metadata for it, hence they can respond to CompleteMultipartUpload() quickly. For minio we decided to keep the backend simple so that when user uploads a large file he will see one large file inside minio's export directory after upload.

**Comment by krishnasrinivas** [comment]
@henryzzhang one question, is your client on the same machine as minio server?

**Comment by henryzzhang** [comment]
My minio server runs in a virtualbox VM with Ubuntu OS.  Client machine is
a windows 10 machine.  They reside in two different physical machines,
which are on the same local network.  Minio server's host machine has SATA
drive.  You probably want to use different machines for client and minio
server.  Try to reduce the network speed, increase file size and/or part
size to see the timeout.

**Comment by krishnasrinivas** [comment]
@henryzzhang just wanted to make sure that client and server are on different machines. Does my previous explanation make sense on why aws-cli does not time out (even with parallel upload of parts) and your client times out?

**Comment by henryzzhang** [comment]
@Krishna Srinivas Actually, my client does exactly the same as you
described on how aws-cli works. I think the timeout will occur if the disk
i/o during concatenation takes too long when the part size is large
enough.  For the very last batch of parts upload, assuming the order is 50,
49, 48, 47, 46 (using your example), when client finished uploading part#
46 followed immediately by the CompleteMultipartUpload(), the server still
needs to spend time to concatenate the last batch, right?  If that time is
longer than 30 seconds, which is Windows WinHttp default receive timeout,
my client will timeout.

On Thu, Nov 10, 2016 at 10:24 PM, Krishna Srinivas <notifications@github.com

**Comment by krishnasrinivas** [comment]
> the server still needs to spend time to concatenate the last batch, right? 

Right

> If that time is longer than 30 seconds, which is Windows WinHttp default receive timeout, my client will timeout.

Ah, 30 seconds is very less, you are right that client timesout when last batch gets appended, can't you increase client timeout?

**Comment by henryzzhang** [comment]
To the client, 30 sec is already a long time.  Even if the winhttp receive
timeout is increased, which could cause side effects to other http calls of
the client, timeout may still occur in some other cases.  From client's
perspective, once all parts have been uploaded, the file upload is
successful, so there is no need to block the MultipartCompletion REST API
until the concatenation is done.  Server should do the concatenation on its
own.  If server crashes during concatenation, it shall resume the work when
it comes back up.  You may keep sdk api CompleteMultipartUpload() as a
blocking call, but it will be nice to make the REST API async.

**Comment by jboelter** [comment]
May want to go this direction to prevent timeouts.  Per the AWS spec: http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadComplete.html

_Processing of a Complete Multipart Upload request could take several minutes to complete. After Amazon S3 begins processing the request, it sends an HTTP response header that specifies a 200 OK response. While processing is in progress, **Amazon S3 periodically sends whitespace characters to keep the connection from timing out**. Because a request could fail after the initial 200 OK response has been sent, it is important that you check the response body to determine whether the request succeeded._

**Comment by harshavardhana** [comment]
> May want to go this direction to prevent timeouts. Per the AWS spec: http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadComplete.html
> 
> Processing of a Complete Multipart Upload request could take several minutes to complete. After Amazon S3 begins processing the request, it sends an HTTP response header that specifies a 200 OK response. While processing is in progress, Amazon S3 periodically sends whitespace characters to keep the connection from timing out. Because a request could fail after the initial 200 OK response has been sent, it is important that you check the response body to determine whether the request succeeded.

If you see the history of the commits we had this feature :-) we removed it on purpose since it became complicated and was not needed for erasure coded backend.  

@krishnasrinivas is working on a fix to address the last batch problem which was seen in @henryzzhang  which should fix the double concatenation that is happening under certain types of workload.

**Comment by GarimaKapoor** [comment]
@krishnasrinivas is this fixed?

**Comment by krishnasrinivas** [comment]
@GarimaKapoor there is some race condition in the fix https://github.com/minio/minio/pull/3304 ... will report back once that race condition is fixed.

**Comment by krishnasrinivas** [comment]
@henryzzhang can you check with the latest release? a bug fix related to this issue has been committed.

**Comment by harshavardhana** [comment]
> @henryzzhang can you check with the latest release? a bug fix related to this issue has been committed.

A related bug fix was made to the minio master and released as well, closing this bug. Please re-open if seen again.

**Comment by saviodsouza** [comment]
I use rclone with minio single disk setup.
Uploads are 1/7 speed compared to Samba transfer.

Is it because of concatenation (joining of file pieces at the backend) it's performance is poor..

**Comment by krishnasrinivas** [comment]
@saviodsouza Can you try with `--s3-upload-cutoff 5000MB` option? Reference: https://github.com/ncw/rclone/blob/master/docs/content/s3.md
Please see if this improves performance (assuming most of your files are less than 5GB). If it improves the performance then it indicates that the background append is causing the perf issue.

**Comment by saviodsouza** [comment]
The S3 cutoff option was not available in rclone sync https://rclone.org/
commands/rclone_sync/ so used something similar s3-chunk-size 2G &
--s3-upload-concurrency 1 which seemed good.

First upload ran really fast  than stuck at 100% for nearly the same amount
of time for upload and completed.

But I'll wait for few days (nothing to sync for now) & sync again to watch
and see.

:)

**Comment by krishnasrinivas** [comment]
Looks like `s3-upload-cutoff` was a recent commit, you can get the rclone for master branch here: https://beta.rclone.org/rclone-beta-latest-linux-amd64.zip

**Comment by krishnasrinivas** [comment]
> so used something similar s3-chunk-size 2G &
--s3-upload-concurrency 1 which seemed good

This won't help as it will still use multipart but with larger chunk size.

**Comment by challarao** [comment]
> If you see the history of the commits we had this feature :-) we removed it on purpose since it became complicated and was not needed for erasure coded backend.

Is it possible to revisit this?

**Comment by krishnasrinivas** [comment]
@challarao this problem is not seen anymore as we do background append as and when the parts get uploaded (and hence completemultipart call is quick). If you are using a reverse proxy you should set sticky session based on source IP affinity.

**Comment by challarao** [comment]
@krishnasrinivas Unfortunately, I'm seeing the request timeout with gcs gateway too. I'm trying to upload 20 gb file. I'm still not sure why since it doesn't need any append with gcs backend. Also it doesn't matter if I use sticky session or not with gcs backend even though if I load balance. Am I right?

With FS backend we don't use multiple minio servers, so lack of sticky session is not an issue.  With FS I didn't see the issue if the bandwidth is typical, but when it is very high -- client in the same cluster I saw 
 the issue.

Since it's hard to determine when there would be a random slowness or if the client will be in a high band width environment, which makes speed greater than disk speed,  I think you should not assume it wouldn't timeout. I think always sending white space characters is a good idea.

**Comment by harshavardhana** [comment]
> @krishnasrinivas Unfortunately, I'm seeing the request timeout with gcs gateway too. I'm trying to upload 20 gb file. I'm still not sure why since it doesn't need any append with gcs backend. Also it doesn't matter if I use sticky session or not with gcs backend even though if I load balance. Am I right?
> 
> With FS backend we don't use multiple minio servers, so lack of sticky session is not an issue. With FS I didn't see the issue if the bandwidth is typical, but when it is very high -- client in the same cluster I saw
> the issue.
> 
> Since it's hard to determine when there would be a random slowness or if the client will be in a high band width environment, which makes speed greater than disk speed, I think you should not assume it wouldn't timeout. I think always sending white space characters is a good idea.

A 50sec timeout is too low @challarao i would suggest it should be close to 15minutes.

**Comment by krishnasrinivas** [comment]
@harshavardhana since the whitespace char fix was simple one, i sent a PR
